fgspectra/models.py

```diff
@@ -1,89 +1,134 @@
-        """Compute the model at frequency and ell combinations.
+    @abstractmethod
+    def __call__(self, *args):
+        """Make it callable."""
+        pass
+
+
+class FactorizedCrossSpectrum(CrossSpectrum):
+    r"""Factorized cross-spectrum
+
+    Cross-spectrum of **one** component for which the scaling in frequency
+    and in multipoles are factorizable
+
+    .. math:: xC_{\ell}^{(ij)} = f(\nu_j) f(\nu_i) C_{\ell}
 
-        Specific models will override this method.
+    Parameters
+    ----------
+    sed : callable
+        :math:`f(\nu)`. It returns an array with shape ``(..., freq)``.
+        It can be :class:`fgspectra.frequency.SED`.
+    cl_args : callable
+        :math:`C_\ell`. It returns an array with shape ``(..., ell)``.
+        It can be :class:`fgspectra.power.PowerSpectrum`
+
+    Note
+    ----
+    The two (optional) sets of extra dimensions ``...`` must be
+    broadcast-compatible.
+    """
+
+    def __init__(self, sed, cl):
+        self._sed = sed
+        self._cl = cl
+
+    def __call__(self, sed_args, cl_args):
+        """Compute the model at frequency and ell combinations.
 
         Parameters
         ----------
-        nu_i : float
-            first frequency channel in cross-spectrum
-        nu_j : float
-            second frequency channel in cross-spectrum
-        ell : 1D array
-            ells at which we evaluate the model
-        **kwargs :
-            additional parameters for the model
+        sed_args : list
+            Arguments for which the `sed` is evaluated.
+        cl_args : list
+            Arguments for which the `cl` is evaluated.
 
         Returns
         -------
-        1D array of floats : model result of same shape as input ells
+        cross : ndarray
+            Cross-spectrum. The shape is ``(..., freq, freq, ell)``.
         """
-        raise NotImplementedError
+        f_nu = self._cl(*sed_args)[..., np.newaxis]
+        return f_nu[..., np.newaxis] * f_nu * self._cl(*cl_args)
+
+
+class CorrelatedFactorizedCrossSpectrum(CrossSpectrum):
+    r"""Factorized cross-spectrum
+
+    Cross-spectrum of multiple correlated components for which the scaling in
+    frequency and in multipoles are factorizable.
+
+    .. math::
 
-    def get_missing(self, input_dict):
-        """Copy self.params and ensure no values are None."""
-        params = self.params.copy()
-        for p in params:
-            if (params[p] is None) and (p in input_dict):
-                params[p] = input_dict[p]
+        xC_\ell^{\nu_i\ \nu_j} = \sum_{kl} f^k(\nu_j) f^n(\nu_i) C^{kn}_\ell
 
-        missing_params = [p for p in params if params[p] is None]
-        if len(missing_params) > 0:
-            raise Exception(f'Missing parameters: {missing_params}.')
+    where :math:`k` and :math:`n` are component indices.
 
-        return params
+    Parameters
+    ----------
+    sed : callable
+        :math:`f(\nu)`. It returns an array with shape ``(comp, ..., freq)``.
```

Comment:
how to organize the shape of SEDs with multiples components?

* `(Comp, …., freq)` natural for accessing the 1st SED with `sed[0]`
* `(…, freq, comp)` natural interpretation as a stack of mixing matrices
* `(…, comp, freq)` similar, but freq is last

Reply:
What do you mean by multiple components? TEB?

Reply:
Suppose you want to use this class to model two correlated power laws. The `comp` index selects the first or the second power law. Makes sense?
Another option is to pass a list of callables and create the array internally

Reply:
Ah right makes sense. Should we go with the first one?

Reply:
I also think it is the most natural thing. Thumbs up